src/device.ts

```diff
@@ -0,0 +1,56 @@
+import { EventEmitter } from "events";
+import { Hub } from "./hub";
+
+import * as Consts from "./consts";
+
+export class Device extends EventEmitter {
+
+    private _hub: Hub;
+    private _portId: number;
+    private _connected: boolean = true;
+    private _type: number;
+    private _busy: boolean = false;
+    private _finished: (() => void) | null = null;
+
+    constructor (hub: Hub, portId: number, type: number = Consts.DeviceType.UNKNOWN) {
+        super();
+        this._hub = hub;
+        this._portId = portId;
+        this._type = type;
+        this.hub.on("detach", (device) => {
+            if (device.portId === this.portId) {
+                this._connected = false;
+                this.emit("detach");
+            }
+        });
+    }
+
+    public get connected () {
+        return this._connected;
+    }
+
+    public get hub () {
+        return this._hub;
+    }
+
+    public get portId () {
+        return this._portId;
+    }
+
+    public get port () {
+        return this.hub.getPortNameForPortId(this.portId);
+    }
+
+    public get type () {
+        return this._type;
+    }
+
+    public send (data: Buffer, characteristic: string = Consts.BLECharacteristic.LPF2_ALL, callback?: () => void) {
+        this.hub.send(data, characteristic, callback);
```

Comment:
Doesn't this send `LPF2_ALL` characteristic messages to WeDo2 hub? As a side note, I am not sure if continued support of the WeDo hub makes sense. The new devices will not be supported on it anyway since the FW is not updatable (according to a Lego designer) and it has limited availability (which also makes it expensive).

Reply:
Right now, yes, eventually I want to have each device method (ie. `setSpeed`) switch characteristics and message based on hub type.

Regarding the WeDo 2.0 support - I've also been thinking about this recently, but I think I want to keep support personally, as I actually use a few WeDo 2.0 hubs myself. :) They are currently the only hub type that offers an AC adapter so it can run forever.

On the topic of firmware support, that's correct - however most newer devices can be supported on it in some way. See the compatibility matrix in the readme of this project.

Eventually I might drop it though - especially if Lego releases a hub with an AC adapter.

Reply:
Yeah, I understand that WeDo support is mostly for people who already own one. Iteresting about the AC adapter though, I have not heard about it. Is this stand alone or the one with rechargeable battery allows to bypass the battery? Long time ago I especially for that purpose bought Lego [RCX with an AC plug](https://www.bricklink.com/v2/catalog/catalogitem.page?P=884a#T=I), even though I already had 2 without :)

Reply:
> Yeah, I understand that WeDo support is mostly for people who already own one. Iteresting about the AC adapter though, I have not heard about it. Is this stand alone or the one with rechargeable battery allows to bypass the battery? Long time ago I especially for that purpose bought Lego [RCX with an AC plug](https://www.bricklink.com/v2/catalog/catalogitem.page?P=884a#T=I), even though I already had 2 without :)

It's the rechargeable battery pack for the WeDo 2.0 hub - it has a built in port to plug in an AC adapter. :)

https://education.lego.com/en-us/products/wedo-2-0-add-on-power-pack-by-lego-education/5004838

That will allow you to keep a WeDo 2.0 hub running forever.

Reply:
I've implemented some WeDo2 support into some devices - it's actually on par with master feature wise, but I may implement some more modes, once I test to see what works and doesn't!

It's much simpler to implement WeDo2 support with each motor/sensor having its own device class.